blobmodel/pulse_shape.py

```diff
@@ -0,0 +1,74 @@
+from abc import ABC, abstractmethod
+import numpy as np
+from nptyping import NDArray
+
+
+class AbstractBlobShape(ABC):
+    """Abstract class containing the blob pulse shapes. Two-dimensional blob
+    pulse shapes are written in the form.
+
+    phi(theta_x, theta_y) = phi_x(theta_x) * phi_y(theta_y).
+    """
+
+    @abstractmethod
+    def get_pulse_shape_prop(self, theta_prop: NDArray, kwargs):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_pulse_shape_perp(self, theta_perp: NDArray, kwargs):
+        raise NotImplementedError
+
+
+class BlobShapeImpl(AbstractBlobShape):
+    """Implementation of the AbstractPulseShape class."""
+
+    __SHAPE_NAMES__ = {"exp", "gauss", "2-exp", "lorentz", "secant"}
+
+    def __init__(self, pulse_shape_prop="gauss", pulse_shape_perp="gauss"):
+        self.__GENERATORS__ = {
+            "exp": BlobShapeImpl._get_exponential_shape,
+            "gauss": BlobShapeImpl._get_gaussian_shape,
+            "2-exp": BlobShapeImpl._get_double_exponential_shape,
+            "lorentz": BlobShapeImpl._get_lorentz_shape,
+            "secant": BlobShapeImpl._get_secant_shape,
+        }
+        self.pulse_shape_prop = pulse_shape_prop
+        self.pulse_shape_perp = pulse_shape_perp
+        if (
+            pulse_shape_prop not in BlobShapeImpl.__SHAPE_NAMES__
+            or pulse_shape_perp not in BlobShapeImpl.__SHAPE_NAMES__
+        ):
+            raise NotImplementedError(
+                f"{self.__class__.__name__}.blob_shape not implemented"
+            )
+
+    def get_pulse_shape_prop(self, theta: np.ndarray, kwargs) -> np.ndarray:
+        return self.__GENERATORS__.get(self.pulse_shape_prop)(theta, kwargs)
+
+    def get_pulse_shape_perp(self, theta: np.ndarray, kwargs) -> np.ndarray:
+        return self.__GENERATORS__.get(self.pulse_shape_perp)(theta, kwargs)
+
+    @staticmethod
+    def _get_exponential_shape(theta: np.ndarray, kwargs) -> np.ndarray:
+        return np.exp(theta) * np.heaviside(-1.0 * theta, 1)
+
+    @staticmethod
+    def _get_lorentz_shape(theta: np.ndarray, kwargs) -> np.ndarray:
+        return 1 / (np.pi * (1 + theta**2))
+
+    @staticmethod
+    def _get_double_exponential_shape(theta: np.ndarray, kwargs) -> np.ndarray:
+        lam = kwargs["lam"]
```

Comment:
So far, `kwargs` is only used for `lam`, are you planning to use it for something else? Otherwise I would think calling it `lam` directly would make it more readable

Reply:
Will be used for other pulse shapes yeah (potentially there could be more than one shape parameter too, although I don't see the use case of that yet)